blockchain/chain_sync/src/consensus.rs

```diff
@@ -64,3 +69,123 @@ pub async fn collect_errs<E>(
         }),
     }
 }
+
+/// The `Proposer` trait expresses the ability to "mine", or in more general,
+/// to propose blocks to the network according to the rules of the consensus
+/// protocol.
+///
+/// It is separate from the `Consensus` trait because it is only expected to
+/// be called once, then left to run in the background and try to publish
+/// blocks to the network which may or may not be adopted.
+///
+/// It exists mostly as a way for us to describe what kind of dependencies
+/// mining processes are expected to take.
+#[async_trait]
+pub trait Proposer {
+    /// Start proposing blocks in the background and never return, unless
+    /// something went horribly wrong. Broadly, they should select messages
+    /// from the mempool, come up with a total ordering for them, then create
+    /// blocks and publish them to the network.
+    ///
+    /// To establish total ordering, the proposer might have to communicate
+    /// with other peers using custom P2P messages, however that is its own
+    /// concern, the dependencies to implement a suitable network later must
+    /// come from somewhere else, because they are not common across all
+    /// consensus variants.
+    async fn run<DB, MP>(
+        self,
+        // NOTE: We will need access to the `ChainStore` as well, or, ideally
```

Comment:
Right, yes, I figured that Forest mimics Lotus in this regard. 

It's just sort of an anti-pattern in Object Oriented programming to expose dependencies and chain together lots of accessors, like `it.a().b().c().do_something(x)` instead of `it.do_abc(x)`, and `manager.chain().database().put(...)` completely obfuscates any ideas about what someone might want to do with `manager`, it's just a hold-all to allow you to do almost anything, because the `ChainStore` is super permissive.

Ideally we could give something like the `MessagePool` gets via the `Provider` interface, or limited views over individual stuff like `ChainStoreReader`. I left it like this because we're already passing the `StateManager` to the `validate_block` function, but it's worth a rethink at some point.